### ISSUE TITLE
Mark operation output and request/response body enums as @frozen 

### ISSUE THREAD

**Issue by PadraigK** [apple/swift-openapi-generator#105]
Since it's not expected that the generated code should support library evolution, we could add @frozen to all of the generated enums. This would greatly improve the ergonomics of switching over enums in the generated code. 

Borrowing an example from [another issue](https://github.com/apple/swift-openapi-generator/issues/104):
```swift
switch result {
case let .ok(reponse):
    case let .json(payload):
        print(payload.body, payload.headers)
    @unknown default:
        break
    }
@unknown default:
    break
} 
```

would become 

```swift
switch result {
case let .ok(reponse):
    case let .json(payload):
        print(payload.body, payload.headers)
    }
} 
```

**Comment by simonjbeaumont** [comment]
@PadraigK thanks for the suggestion.

We do generate `@frozen` enums in a bunch of places already[^0][^1][²] so potentially we might be able to do it for the response body enum too.

@czechboy0 might have some opinions here. Assuming there's no reason not to, is it something you'd want to work on?

[^0]: https://github.com/apple/swift-openapi-generator/blob/ec7a1c5/Tests/OpenAPIGeneratorReferenceTests/Resources/ReferenceSources/Petstore/Types.swift#L94
[^1]: https://github.com/apple/swift-openapi-generator/blob/ec7a1c5/Tests/OpenAPIGeneratorReferenceTests/Resources/ReferenceSources/Petstore/Types.swift#L219
[^2]: https://github.com/apple/swift-openapi-generator/blob/ec7a1c5/Tests/OpenAPIGeneratorReferenceTests/Resources/ReferenceSources/Petstore/Types.swift#L430

**Comment by czechboy0** [comment]
Yeah we should add it to all our enums, I don't think there are any downsides.

**Comment by simonjbeaumont** [comment]
Great, should be a simple change in `translateRequestBodyInTypes`. I'll do it on Monday, unless someone else beats me to it.

**Comment by simonjbeaumont** [comment]
@PadraigK I've opened a PR for this. I retitled the issue to make it a little clearer for posterity, because there's a bunch of enums we generate which don't have any cases (just namespace types).

Thanks again for taking the time to try out the generator and for filing the issue!

**Comment by simonjbeaumont** [comment]
I discussed this some more with some folks and there's some feedback that we might not want to do this.

To summarise, they highlighted that adopters should only be forced to add the `@unknown default` case if _their_ target is being compiled with library evolution and this isn't something we should be encouraging. They also suggest that, since `@frozen` is used for library evolution use cases, using it anywhere makes an implication that we're in the library-evolution game, which we are not.

As I highlighted earlier in this issue thread, we are currently already using `@frozen` in a couple of places. These are for enum (OpenAPI enum) and OneOf types.

I think we should take a consistent approach here. My PR moved us in the direction of consistent application of `@frozen` to all places where we generate a Swift enum with cases. But I think we should also consider whether this was completely wrong, and we should instead be removing it everywhere?

//cc @czechboy0 @FranzBusch @dnadoba

**Comment by dnadoba** [comment]
>Since it's not expected that the generated code should support library evolution, we could add @Frozen to all of the generated enums.

@PadraigK  `@frozen` only has an effect if library evolution mode is turned on:

> When the compiler isn’t in library evolution mode, all structures and enumerations are implicitly frozen, and this attribute is ignored.
> https://docs.swift.org/swift-book/documentation/the-swift-programming-language/attributes/#frozen

Adding `@frozen` is a no-op without library evolution mode turned on and `@unknown default` is not required. Are you sure you haven't turned on library evolution mode accidentally? How do you integrate/build the generated code and the OpenAPI runtime library?

**Comment by czechboy0** [comment]
I'm for adding it everywhere, since it's a no-op to code not using library evolution, and solves a major usability issue for folks using library evolution. It's asymmetrical, so for pragmatic reasons here, I'd do the thing that unblocks a set of adopters, and doesn't hurt other adopters.

That said, we should probably also add a paragraph to our docs about how to use generated code, and mention that the generated code is not designed for having a stable ABI (like we already call out the lack of a stable API).

**Comment by FranzBusch** [comment]
> I'm for adding it everywhere, since it's a no-op to code not using library evolution, and solves a major usability issue for folks using library evolution

I don't think I agree with this. The problem is if we start adding annotations for library evolution such as `@frozen` it conveys the feeling that we are indeed supporting ABI stability in both the generated code but also our runtime library. This is simply not true and would be a significant piece of work. While adding `@frozen` is silencing some of the errors and warnings for people it doesn't remove the danger of ABI mismatches at runtime.

However, adopters do have a way out of this. When they want to ship an ABI stable framework that uses generated code and the runtime they just need to make sure that all imports are marked with `@_implementationOnly`. This should make sure to hide the dependency for them and that they don't use any of the types in their public interface.

**Comment by simonjbeaumont** [comment]
> After talking with @czechboy0 about this offline.

We've not really added any configuration options so far, but this is not the first issue that could justify adding one. I'd be open to us starting to use it more.

**Comment by czechboy0** [comment]
I think we should think about how we could _add_ configuration for cases where users genuinely need both options, in a backwards-compatible way. Here, I'm not sure it's needed yet, as always adding `@frozen` covers both sets of users (as without LE, it's ignored), the only difference is strictness.

As you can imagine, the cost of each configuration option in the generator is adding a whole axis to the test matrix, so it's not cheap, and each of them need to be well justified. I'm not saying this one isn't, just that we should start without it (make the behavior consistent, add `@frozen` to the missing places, since we already add it some places today), and later if we still feel _not_ generating them is a valuable feature, add an opt-in configuration option to skip them.

**Comment by simonjbeaumont** [comment]
So where are we landing on this one? Should we merge #109 as-is (i.e. generating `@frozen` by default, with no configuration option)? //cc @FranzBusch @czechboy0
